### ISSUE TITLE
Can not iterate through directory entries on Android

### ISSUE THREAD

**Issue by alexeikh** [boostorg/filesystem#72]
### Prerequisites

- Android device with Android OS below 6.0, for example Samsung Galaxy S6 with Android 5.0.2. The issue should be reproducible on Android Emulator as well.
- Android NDK with enabled Unified Headers, for example version r16b
- Compiler: Clang (shouldn't be important)
- STL: GNU STL (shouldn't be important)
- Selected (minimum) Android API level: 14 (shouldn't be important)

### How to reproduce

Compile the following sample code as Android executable.
```c++
#include <boost/filesystem.hpp>
#include <iostream>

int main() {
    namespace fs = boost::filesystem;

    fs::directory_iterator iter("/data/local/tmp");
    for (const fs::path& p: iter)
        std::cout << "Directory entry found: " << p.c_str() << std::endl;

    return 0;
}
```

Copy the resulting executable to the device and run it:
```sh
adb push libs/armeabi-v7a/sample-cmdline-app /data/local/tmp/
adb shell /data/local/tmp/sample-cmdline-app
```

### Expected behavior

All directory entries are listed, for example:
```
Directory entry found: /data/local/tmp/sample-cmdline-app
Directory entry found: /data/local/tmp/some-another-file.txt
```

### Actual behavior

`boost::filesystem::filesystem_error` is thrown on the second iteration of the loop:
```
Directory entry found: /data/local/tmp/sample-cmdline-app
terminate called after throwing an instance of 'boost::filesystem::filesystem_error'
  what():  boost::filesystem::directory_iterator::operator++: Function not implemented: "/data/local/tmp"
```

**Comment by alexeikh** [comment]
The cause of this issue seems to be buggy readdir_r() implementation on Android 5.x and below. Applying patch #68 fixes the issue.

**Comment by alexeikh** [comment]
Applying patch #68 fixes the issue.

**Comment by alexeikh** [comment]
The same issue is also reported on the Boost's Trac:
https://svn.boost.org/trac10/ticket/13172

There, another patch is proposed for the fix:
https://gist.github.com/webmaster128/5912a70d100e9ef341df67b177c465d6

**Comment by alexeikh** [comment]
Pull request #51 implements exactly errno resetting in 2 places and also fixes this issue.

**Comment by rcdailey** [comment]
How will we know which version of Boost this fix becomes available in?

**Comment by alexeikh** [comment]
The issue has been fixed by merging #51 in.
